### DISCUSSION TITLE
I search for documentation about the code

### DISCUSSION THREAD

**Discussion by bukru** [SynthstromAudible/DelugeFirmware#2863] [Q&A]
I start to have a look on the code and i'm asking if there is some code documentation such as diagrams etc to help understanding the code architecture ?

I found some text documentation under docs folder. But i'm looking for more if it exists of course.

**Comment by bukru** [comment]
Or any advice about where to start the analyse whould be helpfull for me

**Reply by RJ-Eckie** [reply]
I'm not sure if I'm the best person to respond to this as I've just been getting into the code for 3-4 weeks or so, but I can give you my point of view:

There are very few comments in the code currently and it's pretty difficult to navigate. I've also not seen any guides on where to start! I think refactoring and documenting the code itself could be one of the more important jobs there is currently in the project. I would personally love to make a start with writing some overarching docs, but I don't wanna step on the toes of the people who have been contributing for a very long time, so I've just been familiarizing myself with the codebase by squashing bugs!

Anyhow, having said all that, here's what worked for me:
- I'm picking issues from github that look manageable and try to fix them, which leads me to investigating the code
- Most of the user interface can be found in /src/deluge/gui/views. You'll recognize the main modes there. Song mode is called Session View and holds both rows and grid view. The 7seg/oled display and other physical ui elements are under /src/deluge/hid/. These locations are good places to start
- The entry point in the views is often [control]Action. For example: padAction() (for the main grid), buttonAction(), selectAction(), etc.

Alright, that should give you at least a place to start!

**Reply by seangoodvibes** [reply]
> I'm not sure if I'm the best person to respond to this as I've just been getting into the code for 3-4 weeks or so, but I can give you my point of view:
> 
> There are very few comments in the code currently and it's pretty difficult to navigate. I've also not seen any guides on where to start! I think refactoring and documenting the code itself could be one of the more important jobs there is currently in the project. I would personally love to make a start with writing some overarching docs, but I don't wanna step on the toes of the people who have been contributing for a very long time, so I've just been familiarizing myself with the codebase by squashing bugs!
> 
> Anyhow, having said all that, here's what worked for me:
> - I'm picking issues from github that look manageable and try to fix them, which leads me to investigating the code
> - Most of the user interface can be found in /src/deluge/gui/views. You'll recognize the main modes there. Song mode is called Session View and holds both rows and grid view. The 7seg/oled display and other physical ui elements are under /src/deluge/hid/. These locations are good places to start
> - The entry point in the views is often [control]Action. For example: padAction() (for the main grid), buttonAction(), selectAction(), etc.
> 
> Alright, that should give you at least a place to start!

Hey RJ, really appreciate you jumping onboard and helping us fix bugs and just overall wanting to contribute!

I personally don't think you'd be stepping on anyone's toes by writing the overarching documentation you speak of. If anything you'd provide us with a great starting point to provide feedback and revise further.

Documentation is one area we are lacking in and we are currently building a website to help with that. And while that website is primarily focused on users (eg a manual for the Deluge), we'd love to have documentation for developers as well so that more people can jump onboard and help out.

In the meantime thanks again for your help! Apologies in advance if the PR's don't get approved right away. The team is a bit short on time these days so things are progressing at a different pace than it did in the past.

**Reply by seangoodvibes** [reply]
> I'm not sure if I'm the best person to respond to this as I've just been getting into the code for 3-4 weeks or so, but I can give you my point of view:
> 
> There are very few comments in the code currently and it's pretty difficult to navigate. I've also not seen any guides on where to start! I think refactoring and documenting the code itself could be one of the more important jobs there is currently in the project. I would personally love to make a start with writing some overarching docs, but I don't wanna step on the toes of the people who have been contributing for a very long time, so I've just been familiarizing myself with the codebase by squashing bugs!
> 
> Anyhow, having said all that, here's what worked for me:
> - I'm picking issues from github that look manageable and try to fix them, which leads me to investigating the code
> - Most of the user interface can be found in /src/deluge/gui/views. You'll recognize the main modes there. Song mode is called Session View and holds both rows and grid view. The 7seg/oled display and other physical ui elements are under /src/deluge/hid/. These locations are good places to start
> - The entry point in the views is often [control]Action. For example: padAction() (for the main grid), buttonAction(), selectAction(), etc.
> 
> Alright, that should give you at least a place to start!

You're also welcome to join our discord if you want to chat with the Dev team directly :)

**Reply by RJ-Eckie** [reply]
> Hey RJ, really appreciate you jumping onboard and helping us fix bugs and just overall wanting to contribute!
> 
> I personally don't think you'd be stepping on anyone's toes by writing the overarching documentation you speak of. If anything you'd provide us with a great starting point to provide feedback and revise further.
> 
> Documentation is one area we are lacking in and we are currently building a website to help with that. And while that website is primarily focused on users (eg a manual for the Deluge), we'd love to have documentation for developers as well so that more people can jump onboard and help out.
> 
> In the meantime thanks again for your help! Apologies in advance if the PR's don't get approved right away. The team is a bit short on time these days so things are progressing at a different pace than it did in the past.

Thanks 🙏🏻

I've just gotten a Deluge (again) after selling my first just as the first oled retrofits were starting. I'm really loving it in its current iteration and I'm very excited to be able to fix hiccups and implement new functionality on it!

I'll start a new branch for documenting the code, as well as ideas about where and how I would start refactoring and what coding guidelines would be helpful - and add it as a draft PR. Anyone can read along and jump in with additions if they want to. I haven't really worked on a larger project before, so I don't really have many expectations about how things are handled. I'm also just finding out how the whole github thing works so forgive me any mistakes.

I like being informed on where the project stands, what the plans are etc. But from an adhd/mental health perspective I really try to limit the internet coming at me too directly and all at once. For that reason I avoid Discord as much as possible. A forum-style place like this is much easier to handle.

**Reply by seangoodvibes** [reply]
> > Hey RJ, really appreciate you jumping onboard and helping us fix bugs and just overall wanting to contribute!
> > I personally don't think you'd be stepping on anyone's toes by writing the overarching documentation you speak of. If anything you'd provide us with a great starting point to provide feedback and revise further.
> > Documentation is one area we are lacking in and we are currently building a website to help with that. And while that website is primarily focused on users (eg a manual for the Deluge), we'd love to have documentation for developers as well so that more people can jump onboard and help out.
> > In the meantime thanks again for your help! Apologies in advance if the PR's don't get approved right away. The team is a bit short on time these days so things are progressing at a different pace than it did in the past.
> 
> Thanks 🙏🏻
> 
> I've just gotten a Deluge (again) after selling my first just as the first oled retrofits were starting. I'm really loving it in its current iteration and I'm very excited to be able to fix hiccups and implement new functionality on it!
> 
> I'll start a new branch for documenting the code, as well as ideas about where and how I would start refactoring and what coding guidelines would be helpful - and add it as a draft PR. Anyone can read along and jump in with additions if they want to. I haven't really worked on a larger project before, so I don't really have many expectations about how things are handled. I'm also just finding out how the whole github thing works so forgive me any mistakes.
> 
> I like being informed on where the project stands, what the plans are etc. But from an adhd/mental health perspective I really try to limit the internet coming at me too directly and all at once. For that reason I avoid Discord as much as possible. A forum-style place like this is much easier to handle.

Totally understand regarding discord being too overwhelming. It happens to me too. Chatting here is perfectly fine :) If you have any questions for me or the other dev's please feel free to ask.

Thanks again for your help, we really appreciate it and welcome it.

**Reply by RJ-Eckie** [reply]
> Totally understand regarding discord being too overwhelming. It happens to me too. Chatting here is perfectly fine :) If you have any questions for me or the other dev's please feel free to ask.
> 
> Thanks again for your help, we really appreciate it and welcome it.

We're drifting a little off-topic here, but just one question for now: what are the intentions for 1.2? Are we working towards a somewhat-soon release? I've been focusing on fixing the v1.2 issues, but I'm also seeing new issues being added to the 1.2 roadmap sometimes

**Reply by seangoodvibes** [reply]
> > Totally understand regarding discord being too overwhelming. It happens to me too. Chatting here is perfectly fine :) If you have any questions for me or the other dev's please feel free to ask.
> > Thanks again for your help, we really appreciate it and welcome it.
> 
> We're drifting a little off-topic here, but just one question for now: what are the intentions for 1.2? Are we working towards a somewhat-soon release? I've been focusing on fixing the v1.2 issues, but I'm also seeing new issues being added to the 1.2 roadmap sometimes

Yes we're working towards a 1.2 release.

At this point we're definitely not adding new features / functionality / behaviours to 1.2. The scope is closed.

However as new bugs get identified, new issues are opened. If the issue should fixed as part of 1.2, then we tag it to release 1.2. 

For example:
- the stem export bugs need to be fixed because stem export is a new 1.2 feature and part of the stem export functionality is not working as intended.
- if we identify behaviour in 1.2 that is not functioning as expected (e.g. it worked a certain way and it's not working that way in 1.2 and we didn't intentionally change it, then that behaviour change is probably a bug and we should open an issue and fix it).

So we're working towards closing off all the issues tagged to release 1.2. 

Ultimately 1.2 should perform the same, if not better than 1.2 and it should be an enhancement to 1.1 in all respects.
